### PR TITLE
Feature/battery handling controller

### PR DESCRIPTION
**Pull request by wgerbl**
please review

### CODE REVIEW

**Comment by sfeilmeier** [comment]
You did not mark anybody for review...

File: io.openems.edge.battery.soltaro/src/io/openems/edge/battery/soltaro/single/versionb/SingleRack.java [hunk 1]
```diff
@@ -393,8 +393,32 @@ private void writeValuesInApiChannels() {
...
 		if (minCellVoltageOpt.isPresent()) {
-			int voltageMillivolt = minCellVoltageOpt.get();
-			this.channel(Battery.ChannelId.MIN_CELL_VOLTAGE).setNextValue(voltageMillivolt);
+			int voltage_millivolt = minCellVoltageOpt.get();
+			this.channel(Battery.ChannelId.MIN_CELL_VOLTAGE).setNextValue(voltage_millivolt);
+		}
+		
+		@SuppressWarnings("unchecked")
+		Optional<Integer> maxCellVoltageOpt = (Optional<Integer>) this
+				.channel(SingleRackChannelId.CLUSTER_1_MAX_CELL_VOLTAGE).value().asOptional();
```
Comment (sfeilmeier) [inline io.openems.edge.battery.soltaro/src/io/openems/edge/battery/soltaro/single/versionb/SingleRack.java:402]: Why don't you use direkt mapping in Modbus definition? If you really need "CLUSTER_1_MAX_CELL_VOLTAGE" also to have the same value, Modbus mapping also allows setting two channels from one register

File: io.openems.edge.controller.battery.batteryprotection/bnd.bnd [hunk 2]
```diff
@@ -0,0 +1,15 @@
+Bundle-Name: OpenEMS Edge Controller io.openems.edge.controller.battery.batteryprotection
```
Comment (sfeilmeier) [inline io.openems.edge.controller.battery.batteryprotection/bnd.bnd:1]: Fix Bundle-Name

File: io.openems.edge.controller.battery.batteryprotection/bnd.bnd [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+Bundle-Vendor: OpenEMS Association e.V.
```
Comment (sfeilmeier) [inline io.openems.edge.controller.battery.batteryprotection/bnd.bnd:2]: Bundle-Vendor is "FENECON GmbH"

File: io.openems.edge.controller.battery.batteryprotection/doc/description.txt [hunk 4]
```diff
@@ -0,0 +1,6 @@
+UNDEFINED: 		Undefined status is always reached when one of the values are undefined 
+NORMAL: 		If no limit is reached this is the normal state
+LIMIT:			If voltage or temperature limits are reached, charging or discharging will be denied 
+FORCE_CHARGE:	If cell voltage or SoC is below a certain limit value, charging is forced
+FULL_CHARGE:	If battery is not charged or discharged for a certain time period a full charge has to be done
+CHECK:			If charging was forced, discharging is only allowed if SoC has increased
```
Comment (sfeilmeier) [inline io.openems.edge.controller.battery.batteryprotection/doc/description.txt:6]: This information could be added to readme.adoc file. Alternatively provide a ful State-Machine definition using mermaid.js - see https://github.com/OpenEMS/openems/blob/1353269b3765c16268030bace9a7ee97a866d75b/io.openems.edge.ess.generic/doc/statemachine.md

Ah - just saw you did it using drawio, that is also ok. Levaing this as a comment anyway.

File: io.openems.edge.controller.battery.batteryprotection/readme.adoc [hunk 5]
```diff
@@ -0,0 +1,3 @@
+= io.openems.edge.controller.battery.batteryprotection
```
Comment (sfeilmeier) [inline io.openems.edge.controller.battery.batteryprotection/readme.adoc:1]: Fix Readme title + also add some description about the Controller. This information is used to generate https://openems.github.io/openems.io/openems/latest/edge/controller.html

File: io.openems.edge.controller.battery.batteryprotection/src/io/openems/edge/controller/battery/batteryprotection/state/ForceCharge.java [hunk 6]
```diff
@@ -46,16 +48,20 @@ public State getNextState() {
 			this.startTime = LocalDateTime.now();
 		}
 
-		if (isMinCellVoltageReached() || isChargingTimeOver()) {
+		if ( (isMinCellVoltageReached() && isWarningSoCReached()) || isChargingTimeOver()) {
```
Comment (sfeilmeier) [inline io.openems.edge.controller.battery.batteryprotection/src/io/openems/edge/controller/battery/batteryprotection/state/ForceCharge.java:51]: Apply Eclipse autoformat and Checkstyle

File: io.openems.edge.controller.battery.batteryprotection/src/io/openems/edge/controller/battery/batteryprotection/state/Limit.java [hunk 7]
```diff
@@ -1,32 +1,35 @@
...
-	int highTemperature;
-	long unusedTime;
+	private int warningLowCellVoltage;
+	private int criticalLowCellVoltage;
+	private int criticalHighCellVoltage;
+	private int warningSoC;	
+	private int criticalSoC;
+	private int lowTemperature;
+	private int highTemperature;
+	private long unusedTime;
```
Comment (sfeilmeier) [inline io.openems.edge.controller.battery.batteryprotection/src/io/openems/edge/controller/battery/batteryprotection/state/Limit.java:22]: those could be `final`

File: io.openems.edge.battery.soltaro/src/io/openems/edge/battery/soltaro/single/versionb/SingleRack.java (hunk 1 above)
Comment (wgerbl) [reply io.openems.edge.battery.soltaro/src/io/openems/edge/battery/soltaro/single/versionb/SingleRack.java:402]: grown structures ;)

Comment (wgerbl) [reply io.openems.edge.battery.soltaro/src/io/openems/edge/battery/soltaro/single/versionb/SingleRack.java:402]: i'll move it to define protocol

**Comment by sfeilmeier** [comment]
This is obsolete. Will be replaced by a new approach for a Battery Protection Algorithm.

**Pull request closed without merging**